In [2]:
import pandas as pd
import numpy as np
pd.set_option('display.max_rows', 25)

# Data manipulation & collecting

## Cleaning data from the first Dataset

In [3]:
df = pd.read_csv("./Billboard Hot 100 since 1958.csv")

In [4]:
df.columns

Index(['chart_position', 'chart_date', 'song', 'performer', 'song_id',
       'instance', 'time_on_chart', 'consecutive_weeks', 'previous_week',
       'peak_position', 'worst_position', 'chart_debut', 'chart_url'],
      dtype='object')

In [5]:
df.isnull().sum()

chart_position           0
chart_date               0
song                     0
performer                0
song_id                  0
instance                 0
time_on_chart            0
consecutive_weeks    33667
previous_week        33667
peak_position            0
worst_position           0
chart_debut              0
chart_url                0
dtype: int64

In [6]:
df.shape

(338395, 13)

In [7]:
df['chart_date'] = pd.to_datetime(df['chart_date'], format='%Y-%m-%d')
df.drop(['chart_position'], axis=1)
df.sort_values(['chart_date', 'chart_position'], ascending=[True, True], inplace=True)
df.reset_index(drop=True, inplace=True)
df.head(1000)['chart_date'].value_counts()

chart_date
1958-08-02    100
1958-08-09    100
1958-08-16    100
1958-08-23    100
1958-08-30    100
1958-09-06    100
1958-09-13    100
1958-09-20    100
1958-09-27    100
1958-10-04    100
Name: count, dtype: int64

In [8]:
# Check if every date is the same
df['chart_day'] = df['chart_date'].dt.dayofweek
df['chart_day'] = df['chart_day'].map({
    0: 'Monday',
    1: 'Tuesday',
    2: 'Wednesday',
    3: 'Thursday',
    4: 'Friday',
    5: 'Saturday',
    6: 'Sunday'
})
df = df[['chart_position', 'chart_date', 'chart_day', 'song', 'performer' 
        #,'song_id' #, 'instance', 'time_on_chart', 
       #'consecutive_weeks', 'previous_week',
       #'peak_position', 'worst_position', 'chart_debut'
       , 'chart_url'
       ]]
df

chart_position chart_date chart_day               song  \
0                    1 1958-08-02  Saturday   Poor Little Fool   
1                    2 1958-08-02  Saturday           Patricia   
2                    3 1958-08-02  Saturday      Splish Splash   
3                    4 1958-08-02  Saturday  Hard Headed Woman   
4                    5 1958-08-02  Saturday               When   
...                ...        ...       ...                ...   
338390              96 2023-05-27  Saturday            Forever   
338391              97 2023-05-27  Saturday    Private Landing   
338392              98 2023-05-27  Saturday            I Heard   
338393              99 2023-05-27  Saturday            Sunrise   
338394             100 2023-05-27  Saturday              Happy   

                                           performer  \
0                                       Ricky Nelson   
1                      Perez Prado And His Orchestra   
2                                        Bobby Darin   
3                 Elvis Presley With The Jordanaires   
4                                        Kalin Twins   
...                                              ...   
338390                    Lil Baby Featuring Fridayy   
338391  Don Toliver Featuring Justin Bieber & Future   
338392                    YoungBoy Never Broke Again   
338393                                 Morgan Wallen   
338394                                            NF   

                                                chart_url  
0       https://www.billboard.com/charts/hot-100/1958-...  
1       https://www.billboard.com/charts/hot-100/1958-...  
2       https://www.billboard.com/charts/hot-100/1958-...  
3       https://www.billboard.com/charts/hot-100/1958-...  
4       https://www.billboard.com/charts/hot-100/1958-...  
...                                                   ...  
338390  https://www.billboard.com/charts/hot-100/2023-...  
338391  https://www.billboard.com/charts/hot-100/2023-...  
338392  https://www.billboard.com/charts/hot-100/2023-...  
338393  https://www.billboard.com/charts/hot-100/2023-...  
338394  https://www.billboard.com/charts/hot-100/2023-...  

[338395 rows x 6 columns]

In [9]:
df['chart_day'].value_counts()

chart_day
Saturday    338395
Name: count, dtype: int64

100 % of the entries are reset on Saturday so I'm going to drop it to clean the DataFrame

In [10]:
df.drop('chart_day', axis=1, inplace=True)

In [11]:
df.head(100).song.value_counts()

song
Got A Match?                2
Early In The Morning        2
Over And Over               2
Poor Little Fool            1
High School Confidential    1
                           ..
Angel Baby                  1
Enchanted Island            1
Secretly                    1
Think It Over               1
Judy                        1
Name: count, Length: 97, dtype: int64

Here are some repetition, so I'm finding a way to understand it.

In [12]:
df.head(100).chart_position.value_counts()

chart_position
63     2
50     2
74     2
73     1
72     1
      ..
31     1
30     1
29     1
28     1
100    1
Name: count, Length: 97, dtype: int64

In [13]:
df.head(100)[df['chart_position'].isin([50, 74, 63])].sort_values(by='chart_position')

/var/folders/qk/t7psjtl510s8pvxmhbfn6v780000gn/T/ipykernel_1419/2881002666.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df.head(100)[df['chart_position'].isin([50, 74, 63])].sort_values(by='chart_position')


chart_position chart_date                              song  \
49              50 1958-08-02        Baubles, Bangles And Beads   
50              50 1958-08-02  By The Light Of The Silvery Moon   
62              63 1958-08-02          High School Confidential   
63              63 1958-08-02                             Padre   
73              74 1958-08-02                   Gotta Have Rain   
74              74 1958-08-02              Win Your Love For Me   

                                performer  \
49                   The Kirby Stone Four   
50    Jimmy Bowen with the Rhythm Orchids   
62  Jerry Lee Lewis And His Pumping Piano   
63                             Toni Arden   
73                            Eydie Gorme   
74                              Sam Cooke   

                                            chart_url  
49  https://www.billboard.com/charts/hot-100/1958-...  
50  https://www.billboard.com/charts/hot-100/1958-...  
62  https://www.billboard.com/charts/hot-100/1958-...  
63  https://www.billboard.com/charts/hot-100/1958-...  
73  https://www.billboard.com/charts/hot-100/1958-...  
74  https://www.billboard.com/charts/hot-100/1958-...

In [14]:
# Rechecking the db with that on billboard website
df.at[22, 'chart_url']

'https://www.billboard.com/charts/hot-100/1958-08-02'

In [15]:
df[df['chart_date'] == '1958-08-02']

chart_position chart_date                                 song  \
0                1 1958-08-02                     Poor Little Fool   
1                2 1958-08-02                             Patricia   
2                3 1958-08-02                        Splish Splash   
3                4 1958-08-02                    Hard Headed Woman   
4                5 1958-08-02                                 When   
..             ...        ...                                  ...   
95              96 1958-08-02                        Over And Over   
96              97 1958-08-02                     I Believe In You   
97              98 1958-08-02                      Little Serenade   
98              99 1958-08-02  I'll Get By (As Long As I Have You)   
99             100 1958-08-02                                 Judy   

                             performer  \
0                         Ricky Nelson   
1        Perez Prado And His Orchestra   
2                          Bobby Darin   
3   Elvis Presley With The Jordanaires   
4                          Kalin Twins   
..                                 ...   
95                     Thurston Harris   
96                     Robert & Johnny   
97                   The Ames Brothers   
98                      Billy Williams   
99                     Frankie Vaughan   

                                            chart_url  
0   https://www.billboard.com/charts/hot-100/1958-...  
1   https://www.billboard.com/charts/hot-100/1958-...  
2   https://www.billboard.com/charts/hot-100/1958-...  
3   https://www.billboard.com/charts/hot-100/1958-...  
4   https://www.billboard.com/charts/hot-100/1958-...  
..                                                ...  
95  https://www.billboard.com/charts/hot-100/1958-...  
96  https://www.billboard.com/charts/hot-100/1958-...  
97  https://www.billboard.com/charts/hot-100/1958-...  
98  https://www.billboard.com/charts/hot-100/1958-...  
99  https://www.billboard.com/charts/hot-100/1958-...  

[100 rows x 5 columns]

The df is identical to the website

Turns out, Billboard allows two songs to have the same position on a weekly chart.

![image.png](./images/img1.png)

Now, it's time to fix some weeks that have more/less than 100 entries

In [16]:
df.drop('chart_url', axis=1, inplace=True)
df.chart_date.value_counts().value_counts()

count
100    3377
99        5
200       1
Name: count, dtype: int64

In [17]:
df['chart_date'].value_counts()[df['chart_date'].value_counts() == 99]

chart_date
1991-11-16    99
1991-11-23    99
1991-11-30    99
1991-12-07    99
1991-12-14    99
Name: count, dtype: int64

In [18]:
missing_value_dates = ['1991-11-16', '1991-11-23', '1991-11-30', '1991-12-07', '1991-12-14']
df[df['chart_date'] == '1991-12-14']

chart_position chart_date                                      song  \
174096               1 1991-12-14                            Black Or White   
174097               2 1991-12-14  It's So Hard To Say Goodbye To Yesterday   
174098               3 1991-12-14                Set Adrift On Memory Bliss   
174099               4 1991-12-14                  When A Man Loves A Woman   
174100               5 1991-12-14                                All 4 Love   
...                ...        ...                                       ...   
174190              95 1991-12-14                                  Groovin'   
174191              96 1991-12-14                                      Lies   
174192              97 1991-12-14                       Monsters And Angels   
174193              99 1991-12-14                         Walk Through Fire   
174194             100 1991-12-14                         Move Any Mountain   

                   performer  
174096       Michael Jackson  
174097           Boyz II Men  
174098             P.M. Dawn  
174099        Michael Bolton  
174100         Color Me Badd  
...                      ...  
174190                  UB40  
174191                   EMF  
174192  Voice Of The Beehive  
174193           Bad Company  
174194            The Shamen  

[99 rows x 4 columns]

In [19]:
# Finding the missing position on the chart by comparing it to a range from 1..100

missing_pos = [i for i in [*range(1,100)] if i not in
            [*df[df['chart_date'] == '1991-11-16']['chart_position'].values]]
# Finding the index of the song that has the position on the chart right before the missing row to insert
previous_missing_pos = missing_pos[0] - 1
df.query("chart_date.isin(@missing_value_dates) & chart_position == @previous_missing_pos")

chart_position chart_date                     song        performer
173797              98 1991-11-16  Try A Little Tenderness  The Commitments
173895              98 1991-11-23                 Groovin'             UB40
173994              98 1991-11-30           True Companion        Marc Cohn
174093              98 1991-12-07  My Heart Belongs To You       Russ Irwin

In [20]:
# It seems like the first 4 dates in ['1991-11-16', '1991-11-23', '1991-11-30', '1991-12-07', '1991-12-14'] is missing 
# the entry at the same position except for the '1991-12-14'
missing_pos_last = [i for i in [*range(1,100)] if i not in
            [*df[df['chart_date'] == '1991-12-14']['chart_position'].values]]
previous_missing_pos_last = missing_pos_last[0] - 1
df.query("chart_date == '1991-12-14' & chart_position == @previous_missing_pos_last")

chart_position chart_date                 song             performer
174192              97 1991-12-14  Monsters And Angels  Voice Of The Beehive

Adding missing values

In [21]:
from datetime import datetime
df.loc[173797.5] = [99,'1991-11-16','Into The Great Wide Open', 'Tom Petty And The Heartbreakers']
df.loc[173895.5] = [99,'1991-11-23','Ain\'t No Future In Yo\' Fronting', 'M.C. Breed & D.F.C.']
df.loc[174093.5] = [99,'1991-11-30','Stand By My Woman', 'Lenny Kravitz']
df.loc[173994.5] = [99,'1991-12-07','Do You Feel Like I Feel?', 'Belinda Carlisle']
df.loc[174192.5] = [98,'1991-12-14','Into The Great Wide Open', 'Tom Petty And The Heartbreakers']
df['chart_date'] = pd.to_datetime(df['chart_date'], format='%Y-%m-%d')
df = df.sort_index().reset_index(drop=True)


In [22]:
# Recheck
df[df['chart_date'] == '1991-11-16']

chart_position chart_date                                      song  \
173700               1 1991-11-16                                     Cream   
173701               2 1991-11-16          Can't Stop This Thing We Started   
173702               3 1991-11-16                  When A Man Loves A Woman   
173703               4 1991-11-16  It's So Hard To Say Goodbye To Yesterday   
173704               5 1991-11-16                Set Adrift On Memory Bliss   
...                ...        ...                                       ...   
173795              96 1991-11-16                            Love Me All Up   
173796              97 1991-11-16                 Love. . .Thy Will Be Done   
173797              98 1991-11-16                   Try A Little Tenderness   
173798              99 1991-11-16                  Into The Great Wide Open   
173799             100 1991-11-16                              Forever More   

                              performer  
173700            Prince And The N.P.G.  
173701                      Bryan Adams  
173702                   Michael Bolton  
173703                      Boyz II Men  
173704                        P.M. Dawn  
...                                 ...  
173795                       Stacy Earl  
173796                          Martika  
173797                  The Commitments  
173798  Tom Petty And The Heartbreakers  
173799                         Stevie B  

[100 rows x 4 columns]

In [23]:
df['chart_date'].value_counts()[df['chart_date'].value_counts() == 200]

chart_date
2023-01-07    200
Name: count, dtype: int64

In [24]:
df[df['chart_date'] == '2023-01-07']

chart_position chart_date                               song  \
336200               1 2023-01-07    All I Want For Christmas Is You   
336201               1 2023-01-07    All I Want For Christmas Is You   
336202               2 2023-01-07  Rockin' Around The Christmas Tree   
336203               2 2023-01-07  Rockin' Around The Christmas Tree   
336204               3 2023-01-07                   Jingle Bell Rock   
...                ...        ...                                ...   
336395              98 2023-01-07                          Forget Me   
336396              99 2023-01-07                           La Jumpa   
336397              99 2023-01-07                           La Jumpa   
336398             100 2023-01-07                          Gone Girl   
336399             100 2023-01-07                          Gone Girl   

                   performer  
336200          Mariah Carey  
336201          Mariah Carey  
336202            Brenda Lee  
336203            Brenda Lee  
336204           Bobby Helms  
...                      ...  
336395         Lewis Capaldi  
336396  Arcangel & Bad Bunny  
336397  Arcangel & Bad Bunny  
336398                   SZA  
336399                   SZA  

[200 rows x 4 columns]

The week of 2023-01-07 is repeated, that's why there are 200 entries in this week so we'll eliminate the even indices.

In [25]:
for i in range(336200,336399):
    if i % 2 == 0:
        df.drop(i, inplace=True)

In [26]:
df = df.sort_index().reset_index(drop=True)

In [27]:
df

chart_position chart_date               song  \
0                    1 1958-08-02   Poor Little Fool   
1                    2 1958-08-02           Patricia   
2                    3 1958-08-02      Splish Splash   
3                    4 1958-08-02  Hard Headed Woman   
4                    5 1958-08-02               When   
...                ...        ...                ...   
338295              96 2023-05-27            Forever   
338296              97 2023-05-27    Private Landing   
338297              98 2023-05-27            I Heard   
338298              99 2023-05-27            Sunrise   
338299             100 2023-05-27              Happy   

                                           performer  
0                                       Ricky Nelson  
1                      Perez Prado And His Orchestra  
2                                        Bobby Darin  
3                 Elvis Presley With The Jordanaires  
4                                        Kalin Twins  
...                                              ...  
338295                    Lil Baby Featuring Fridayy  
338296  Don Toliver Featuring Justin Bieber & Future  
338297                    YoungBoy Never Broke Again  
338298                                 Morgan Wallen  
338299                                            NF  

[338300 rows x 4 columns]

## add audio features from Spotify API

Todo:
* Find ID of songs and add it to the DF
* Using ID, find audio features of a song
* Add both ID and audio features to the DF

### Finding song ID

Since there are 338300 rows in this DF, finding ID for each row is computationally ineffective since there are a number of duplicates. So the goal here is to just find the unique songs' IDs only.
Assumption: Any song that shares the same name is the same song.

In [28]:
unique_songs = df.song.unique()
len(unique_songs)

25372

Even with unique values, there are *25372* distinct entries that we'll have to find the ID for. Since we are relying on the API to get our information, there will be inherent uncertainties and/or errors (e.g. lost connection, API limit is reached, null result). To ensure the data collecting process will go on smoothly, we'll have to split this array into smaller chunks. If an error occurs or if you want to pause the process, you can terminate it and only have to restart at the last completed chunk.

In [29]:
splitted_unique_songs = np.array_split(unique_songs, 100)
len(splitted_unique_songs[0])

254

There are 254 songs in each array with the exception of the last array. There are 100 arrays in total

### Getting IDs from Spotify API

There is still a scenario in which the song is not on Spotify Search, in this case we'll return NaN instead of pausing the entire process.

In [30]:
import os.path as path
import json
import requests
import spotify_get_cre
import time
if path.isfile('./id_out/99_99.json'):
    print('The process has finished running. Skipped.')
else:

    # Your Spotify API access token
    access_token = spotify_get_cre.get_token()

    # Create a dictionary for song name and id and save it as json to remove the dependency upon the API
    unique_id = {}
    # This is the last position, if you want to resume from a specific point, 
    # change this variable to a specific index from 0..99
    count = 99
    for unique in splitted_unique_songs[count:100]:
        no = 0
        for song in unique:
            
            try:
                # The query parameters for the search
                query_params = {
                    'q': f"{song}",
                    'type': 'track',
                    'limit': 1
                }

                # The Spotify API endpoint for searching
                search_url = 'https://api.spotify.com/v1/search'

                # Define the headers with the access token
                headers = {
                    'Authorization': f'Bearer {access_token}'
                }

                # Make the GET request to the Spotify API
                response = requests.get(search_url, params=query_params, headers=headers)
                result = json.loads(response.content)
                sp_id = result['tracks']['items'][0]['id']
                unique_id[f'{song}'] = sp_id
                print(f'{no} song in {count} added {song}')
                no = no + 1
            except IndexError:
                unique_id[f'{song}'] = 'NaN'
        with open(f"./id_out/{count}_99.json", "w") as outfile:
            json.dump(unique_id, outfile)
        print('Pausing 3 seconds')
        time.sleep(3) # To avoid API bottlenecking
        unique_id = {}
        count = count + 1

The process has finished running. Skipped.


Now, we'll have to merge every ID json files into one

... and it will still only be run once if the file hadn't existed already

In [31]:
if path.isfile('output_id.json'):
    print('The process has finished running. Skipped.')
else:
    import glob
    result = []

    # Define the file paths of the JSON files to merge
    file_paths = glob.glob("./id_out/*.json")
    merged_data = {}

    # Iterate through the file paths
    for file_path in file_paths:
        # Open and read each JSON file
        with open(file_path, 'r') as file:
            data = json.load(file)
        # Merge the data into the merged_data dictionary
        merged_data.update(data)
    # Define the output file path
    output_file = "output_id.json"
    # Write the merged data to the output file
    with open(output_file, 'w') as outfile:
        json.dump(merged_data, outfile, indent=4)

The process has finished running. Skipped.


In [32]:
with open('output_id.json', 'r') as file:
    song_id = json.load(file)
len(song_id)

25372

Add ID column

Save DataFrame to a csv to prevent this algorithm from running everytime

In [33]:
import os.path as path
df['id'] = ''
if path.isfile('billboard_with_id.csv'):
    df = pd.read_csv('billboard_with_id.csv')
else:
    for index, row in df.iterrows():
        df.loc[index, 'id'] = song_id[row['song']]
    df.to_csv('billboard_with_id.csv', index=False)
df

chart_position  chart_date               song  \
0                    1  1958-08-02   Poor Little Fool   
1                    2  1958-08-02           Patricia   
2                    3  1958-08-02      Splish Splash   
3                    4  1958-08-02  Hard Headed Woman   
4                    5  1958-08-02               When   
...                ...         ...                ...   
338295              96  2023-05-27            Forever   
338296              97  2023-05-27    Private Landing   
338297              98  2023-05-27            I Heard   
338298              99  2023-05-27            Sunrise   
338299             100  2023-05-27              Happy   

                                           performer                      id  
0                                       Ricky Nelson  5ayybTSXNwcarDtxQKqvWX  
1                      Perez Prado And His Orchestra  6nmr64Mo4RD3R1pLshvmoG  
2                                        Bobby Darin  40fD7ct05FvQHLdQTgJelG  
3                 Elvis Presley With The Jordanaires  3AlUM4CsiOwVl5Gwxfgwcw  
4                                        Kalin Twins  1mMLMZYXkMueg65jRRWG1l  
...                                              ...                     ...  
338295                    Lil Baby Featuring Fridayy  6IBcOGPsniK3Pso1wHIhew  
338296  Don Toliver Featuring Justin Bieber & Future  52NGJPcLUzQq5w7uv4e5gf  
338297                    YoungBoy Never Broke Again  1tqT6DhmsrtQgyCKUwotiw  
338298                                 Morgan Wallen  0JtgoDDf0ggWaoXCGhcOoT  
338299                                            NF  3ZEno9fORwMA1HPecdLi0R  

[338300 rows x 5 columns]

add audio features columns

Sample audio_feature

reference: [link](https://developer.spotify.com/documentation/web-api/reference/get-several-audio-features)

In [34]:
"""
{
  "audio_features": [
    {
      "acousticness": 0.00242,
      "analysis_url": "https://api.spotify.com/v1/audio-analysis/2takcwOaAZWiXQijPHIx7B",
      "danceability": 0.585,
      "duration_ms": 237040,
      "energy": 0.842,
      "id": "2takcwOaAZWiXQijPHIx7B",
      "instrumentalness": 0.00686,
      "key": 9,
      "liveness": 0.0866,
      "loudness": -5.883,
      "mode": 0,
      "speechiness": 0.0556,
      "tempo": 118.211,
      "time_signature": 4,
      "track_href": "https://api.spotify.com/v1/tracks/2takcwOaAZWiXQijPHIx7B",
      "type": "audio_features",
      "uri": "spotify:track:2takcwOaAZWiXQijPHIx7B",
      "valence": 0.428
    }
  ]
}
"""

'\n{\n  "audio_features": [\n    {\n      "acousticness": 0.00242,\n      "analysis_url": "https://api.spotify.com/v1/audio-analysis/2takcwOaAZWiXQijPHIx7B",\n      "danceability": 0.585,\n      "duration_ms": 237040,\n      "energy": 0.842,\n      "id": "2takcwOaAZWiXQijPHIx7B",\n      "instrumentalness": 0.00686,\n      "key": 9,\n      "liveness": 0.0866,\n      "loudness": -5.883,\n      "mode": 0,\n      "speechiness": 0.0556,\n      "tempo": 118.211,\n      "time_signature": 4,\n      "track_href": "https://api.spotify.com/v1/tracks/2takcwOaAZWiXQijPHIx7B",\n      "type": "audio_features",\n      "uri": "spotify:track:2takcwOaAZWiXQijPHIx7B",\n      "valence": 0.428\n    }\n  ]\n}\n'

In [35]:
if path.isfile('./audio_features/au_99_99.json'):
    print('The process has finished running. Skipped.')
else:
    # Your Spotify API access token
    access_token = spotify_get_cre.get_token()

    # This is the last position, if you want to resume from a specific point, 
    # change this variable to a specific index from 0..99
    count = 5
    for unique in splitted_unique_songs[count:100]:
        no = 0
        ids_part_1 = [] # from 0-99
        ids_part_2 = [] # from 100-199
        ids_part_3 = [] # from 200-254
        merged_data = []
        for index, song in enumerate(unique):
            if index < 100:
                ids_part_1.append(song_id[song])
            elif 100<= index and index < 200:
                ids_part_2.append(song_id[song])
            elif 200 <= index:
                ids_part_3.append(song_id[song])
        ids_part_1 = ','.join(ids_part_1)
        ids_part_2 = ','.join(ids_part_2)
        ids_part_3 = ','.join(ids_part_3)
        ids = [ids_part_1, ids_part_2, ids_part_3]
        for i in ids:
            # The query parameters for the search
            query_params = {
                'ids': f"{i}",
            }

            # The Spotify API endpoint for searching
            search_url = 'https://api.spotify.com/v1/audio-features'

            # Define the headers with the access token
            headers = {
                'Authorization': f'Bearer {access_token}'
            }

            # Make the GET request to the Spotify API
            response = requests.get(search_url, params=query_params, headers=headers)
            result = json.loads(response.content)
            merged_data.extend(result['audio_features'])        
        print(f'Saved au_{count}_99.json')
        with open(f'audio_features/au_{count}_99.json', 'w') as file:
                    json.dump(merged_data, file, indent=4)
        count = count + 1


The process has finished running. Skipped.


Merge into audio_features.json

In [36]:
if path.isfile('audio_featuress.json'):
    print('The process has finished running. Skipped.')
else:
    import glob
    result = []

    # Define the file paths of the JSON files to merge
    file_paths = glob.glob("./audio_features/*.json")
    merged_data = []
    # Iterate through the file paths
    for file_path in file_paths:
        # Open and read each JSON file
        with open(file_path, 'r') as file:
            data = json.load(file)
            # Merge the data into the merged_data dictionary
            merged_data.extend(data)
    # Turn ids into keys
    num_merged_data = {}
    for index, track in enumerate(merged_data):
        if track is not None:
            num_merged_data[f"{track['id']}"] = track
        else:
            print(f'None at {index}')
    output_file = "audio_features.json"
    # Write the merged data to the output file
    with open(output_file, 'w') as outfile:
        json.dump(num_merged_data, outfile, indent=4)
len(num_merged_data)

None at 1729
None at 5630
None at 5945
None at 6081
None at 6871
None at 8033
None at 10946
None at 13976
None at 16718
None at 16839
None at 17373
None at 18527
None at 18789
None at 23366
None at 24601


23855

Add audio features to DF

In [40]:
if path.isfile('DB_Billboard_SP_final.csv'):
    print('The process has finished running. Skipped.')
    df = pd.read_csv('DB_Billboard_SP_final.csv')
else:
    df = df.assign(acousticness='', danceability = '', duration_ms = '', energy = '', instrumentalness = '', key = '', liveness = '',
                loudness = '', mode = '', speechiness = '', tempo = '', time_signature = '', valence = '')
    for index, row in df.iterrows():
        try:
            df.loc[index, 'acousticness'] = num_merged_data[f"{row['id']}"]['acousticness']
            df.loc[index, 'danceability'] = num_merged_data[f"{row['id']}"]['danceability']
            df.loc[index, 'duration_ms'] = num_merged_data[f"{row['id']}"]['duration_ms']
            df.loc[index, 'energy'] = num_merged_data[f"{row['id']}"]['energy']
            df.loc[index, 'instrumentalness'] = num_merged_data[f"{row['id']}"]['instrumentalness']
            df.loc[index, 'key'] = num_merged_data[f"{row['id']}"]['key']
            df.loc[index, 'liveness'] = num_merged_data[f"{row['id']}"]['liveness']
            df.loc[index, 'loudness'] = num_merged_data[f"{row['id']}"]['loudness']
            df.loc[index, 'mode'] = num_merged_data[f"{row['id']}"]['mode']
            df.loc[index, 'speechiness'] = num_merged_data[f"{row['id']}"]['speechiness']
            df.loc[index, 'tempo'] = num_merged_data[f"{row['id']}"]['tempo']
            df.loc[index, 'time_signature'] = num_merged_data[f"{row['id']}"]['time_signature']
            df.loc[index, 'valence'] = num_merged_data[f"{row['id']}"]['valence']
        except KeyError:
            print(f'Can\'t find ID in audio_features at {index}')
    df.to_csv('DB_Billboard_SP_final.csv', index=False)
df

The process has finished running. Skipped.


chart_position  chart_date               song  \
0                    1  1958-08-02   Poor Little Fool   
1                    2  1958-08-02           Patricia   
2                    3  1958-08-02      Splish Splash   
3                    4  1958-08-02  Hard Headed Woman   
4                    5  1958-08-02               When   
...                ...         ...                ...   
338295              96  2023-05-27            Forever   
338296              97  2023-05-27    Private Landing   
338297              98  2023-05-27            I Heard   
338298              99  2023-05-27            Sunrise   
338299             100  2023-05-27              Happy   

                                           performer                      id  \
0                                       Ricky Nelson  5ayybTSXNwcarDtxQKqvWX   
1                      Perez Prado And His Orchestra  6nmr64Mo4RD3R1pLshvmoG   
2                                        Bobby Darin  40fD7ct05FvQHLdQTgJelG   
3                 Elvis Presley With The Jordanaires  3AlUM4CsiOwVl5Gwxfgwcw   
4                                        Kalin Twins  1mMLMZYXkMueg65jRRWG1l   
...                                              ...                     ...   
338295                    Lil Baby Featuring Fridayy  6IBcOGPsniK3Pso1wHIhew   
338296  Don Toliver Featuring Justin Bieber & Future  52NGJPcLUzQq5w7uv4e5gf   
338297                    YoungBoy Never Broke Again  1tqT6DhmsrtQgyCKUwotiw   
338298                                 Morgan Wallen  0JtgoDDf0ggWaoXCGhcOoT   
338299                                            NF  3ZEno9fORwMA1HPecdLi0R   

        acousticness  danceability  duration_ms  energy  instrumentalness  \
0             0.6480         0.474     153933.0   0.338          0.000000   
1             0.5280         0.507     217413.0   0.652          0.000501   
2             0.3850         0.645     131720.0   0.943          0.000000   
3             0.3550         0.477     226973.0   0.307          0.000072   
4             0.0130         0.551     240760.0   0.801          0.000006   
...              ...           ...          ...     ...               ...   
338295        0.1910         0.487     232533.0   0.650          0.000000   
338296        0.0846         0.843     238253.0   0.669          0.000013   
338297        0.2720         0.711     194293.0   0.362          0.000000   
338298        0.2590         0.613     182522.0   0.765          0.000004   
338299        0.3060         0.732     242691.0   0.855          0.000000   

        key  liveness  loudness  mode  speechiness    tempo  time_signature  \
0       0.0    0.1300   -11.528   1.0       0.0299  154.596             4.0   
1       9.0    0.1080    -5.872   0.0       0.0469  118.003             4.0   
2       0.0    0.3700    -1.526   1.0       0.0393  147.768             4.0   
3       0.0    0.0870   -12.476   1.0       0.0336  146.692             4.0   
4       6.0    0.2650    -5.069   1.0       0.0303  128.027             4.0   
...     ...       ...       ...   ...          ...      ...             ...   
338295  0.0    0.0933    -5.195   1.0       0.0253  151.964             4.0   
338296  1.0    0.1050    -4.105   0.0       0.0639  136.979             4.0   
338297  3.0    0.1270   -10.851   0.0       0.0381  117.287             4.0   
338298  2.0    0.0832    -5.058   0.0       0.0318  139.993             4.0   
338299  7.0    0.1200    -4.064   1.0       0.0415  106.004             4.0   

        valence  
0         0.810  
1         0.191  
2         0.965  
3         0.387  
4         0.625  
...         ...  
338295    0.456  
338296    0.435  
338297    0.616  
338298    0.343  
338299    0.221  

[338300 rows x 18 columns]

Convert these columns from objects to numbers

In [42]:
df['acousticness'] = pd.to_numeric(df['acousticness'], errors='coerce')
df['danceability'] = pd.to_numeric(df['acousticness'], errors='coerce')
df['duration_ms'] = pd.to_numeric(df['acousticness'], errors='coerce')
df['energy'] = pd.to_numeric(df['acousticness'], errors='coerce')
df['instrumentalness'] = pd.to_numeric(df['acousticness'], errors='coerce')
df['key'] = pd.to_numeric(df['acousticness'], errors='coerce')
df['liveness'] = pd.to_numeric(df['acousticness'], errors='coerce')
df['loudness'] = pd.to_numeric(df['acousticness'], errors='coerce')
df['mode'] = pd.to_numeric(df['acousticness'], errors='coerce')
df['speechiness'] = pd.to_numeric(df['acousticness'], errors='coerce')
df['tempo'] = pd.to_numeric(df['acousticness'], errors='coerce')
df['time_signature'] = pd.to_numeric(df['acousticness'], errors='coerce')
df['valence'] = pd.to_numeric(df['acousticness'], errors='coerce')

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 338300 entries, 0 to 338299
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   chart_position    338300 non-null  int64  
 1   chart_date        338300 non-null  object 
 2   song              338300 non-null  object 
 3   performer         338300 non-null  object 
 4   id                338258 non-null  object 
 5   acousticness      338146 non-null  float64
 6   danceability      338146 non-null  float64
 7   duration_ms       338146 non-null  float64
 8   energy            338146 non-null  float64
 9   instrumentalness  338146 non-null  float64
 10  key               338146 non-null  float64
 11  liveness          338146 non-null  float64
 12  loudness          338146 non-null  float64
 13  mode              338146 non-null  float64
 14  speechiness       338146 non-null  float64
 15  tempo             338146 non-null  float64
 16  time_signature    33

## Cleaning data again from the final DS

In [89]:
df = df[~df['acousticness'].isnull()]
df

chart_position  chart_date               song  \
0                    1  1958-08-02   Poor Little Fool   
1                    2  1958-08-02           Patricia   
2                    3  1958-08-02      Splish Splash   
3                    4  1958-08-02  Hard Headed Woman   
4                    5  1958-08-02               When   
...                ...         ...                ...   
338295              96  2023-05-27            Forever   
338296              97  2023-05-27    Private Landing   
338297              98  2023-05-27            I Heard   
338298              99  2023-05-27            Sunrise   
338299             100  2023-05-27              Happy   

                                           performer                      id  \
0                                       Ricky Nelson  5ayybTSXNwcarDtxQKqvWX   
1                      Perez Prado And His Orchestra  6nmr64Mo4RD3R1pLshvmoG   
2                                        Bobby Darin  40fD7ct05FvQHLdQTgJelG   
3                 Elvis Presley With The Jordanaires  3AlUM4CsiOwVl5Gwxfgwcw   
4                                        Kalin Twins  1mMLMZYXkMueg65jRRWG1l   
...                                              ...                     ...   
338295                    Lil Baby Featuring Fridayy  6IBcOGPsniK3Pso1wHIhew   
338296  Don Toliver Featuring Justin Bieber & Future  52NGJPcLUzQq5w7uv4e5gf   
338297                    YoungBoy Never Broke Again  1tqT6DhmsrtQgyCKUwotiw   
338298                                 Morgan Wallen  0JtgoDDf0ggWaoXCGhcOoT   
338299                                            NF  3ZEno9fORwMA1HPecdLi0R   

        acousticness  danceability  duration_ms  energy  instrumentalness  \
0             0.6480        0.6480       0.6480  0.6480            0.6480   
1             0.5280        0.5280       0.5280  0.5280            0.5280   
2             0.3850        0.3850       0.3850  0.3850            0.3850   
3             0.3550        0.3550       0.3550  0.3550            0.3550   
4             0.0130        0.0130       0.0130  0.0130            0.0130   
...              ...           ...          ...     ...               ...   
338295        0.1910        0.1910       0.1910  0.1910            0.1910   
338296        0.0846        0.0846       0.0846  0.0846            0.0846   
338297        0.2720        0.2720       0.2720  0.2720            0.2720   
338298        0.2590        0.2590       0.2590  0.2590            0.2590   
338299        0.3060        0.3060       0.3060  0.3060            0.3060   

           key  liveness  loudness    mode  speechiness   tempo  \
0       0.6480    0.6480    0.6480  0.6480       0.6480  0.6480   
1       0.5280    0.5280    0.5280  0.5280       0.5280  0.5280   
2       0.3850    0.3850    0.3850  0.3850       0.3850  0.3850   
3       0.3550    0.3550    0.3550  0.3550       0.3550  0.3550   
4       0.0130    0.0130    0.0130  0.0130       0.0130  0.0130   
...        ...       ...       ...     ...          ...     ...   
338295  0.1910    0.1910    0.1910  0.1910       0.1910  0.1910   
338296  0.0846    0.0846    0.0846  0.0846       0.0846  0.0846   
338297  0.2720    0.2720    0.2720  0.2720       0.2720  0.2720   
338298  0.2590    0.2590    0.2590  0.2590       0.2590  0.2590   
338299  0.3060    0.3060    0.3060  0.3060       0.3060  0.3060   

        time_signature  valence  
0               0.6480   0.6480  
1               0.5280   0.5280  
2               0.3850   0.3850  
3               0.3550   0.3550  
4               0.0130   0.0130  
...                ...      ...  
338295          0.1910   0.1910  
338296          0.0846   0.0846  
338297          0.2720   0.2720  
338298          0.2590   0.2590  
338299          0.3060   0.3060  

[338146 rows x 18 columns]

In [90]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 338146 entries, 0 to 338299
Data columns (total 18 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   chart_position    338146 non-null  int64  
 1   chart_date        338146 non-null  object 
 2   song              338146 non-null  object 
 3   performer         338146 non-null  object 
 4   id                338146 non-null  object 
 5   acousticness      338146 non-null  float64
 6   danceability      338146 non-null  float64
 7   duration_ms       338146 non-null  float64
 8   energy            338146 non-null  float64
 9   instrumentalness  338146 non-null  float64
 10  key               338146 non-null  float64
 11  liveness          338146 non-null  float64
 12  loudness          338146 non-null  float64
 13  mode              338146 non-null  float64
 14  speechiness       338146 non-null  float64
 15  tempo             338146 non-null  float64
 16  time_signature    338146 

# EDA

We'll take
* acousticness
* danceability
* duration_ms
* energy
* instrumentalness
* loudness
* mode
* tempo
* valence

Notwithstanding, we're still going to extract all the data for the sake of API usage, for any future uses.

# EDA

TODO:
* Histogram
